<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Simulate DNA under the SELAC model and evolving rates — SelacSimulatorEvolvingRates • selac</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">selac</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate DNA under the SELAC model and evolving rates</h1>
    </div>

    
    <p>Simulates nucleotide data based on parameters under the SELAC model but assumes either Phi or Ne evolves along the tree.</p>
    

    <pre><span class='fu'>SelacSimulatorEvolvingRates</span>(<span class='no'>phy</span>, <span class='no'>pars</span>, <span class='no'>aa.optim_array</span>, <span class='no'>root.codon.frequencies</span>,
  <span class='kw'>numcode</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>aa.properties</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>nuc.model</span>, <span class='kw'>k.levels</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>diploid</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>pars.to.evolve</span> <span class='kw'>=</span> <span class='st'>"phi"</span>, <span class='kw'>evolve.type</span> <span class='kw'>=</span> <span class='st'>"BM"</span>,
  <span class='kw'>evolve.pars</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>1</span>, <span class='fl'>0</span>), <span class='kw'>Ne.vals.evolved</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>phy</th>
      <td><p>The phylogenetic tree with branch lengths.</p></td>
    </tr>
    <tr>
      <th>pars</th>
      <td><p>A vector of parameters used for the simulation. They are ordered as follows: C.q.phi, alpha, beta, and Ne.</p></td>
    </tr>
    <tr>
      <th>aa.optim_array</th>
      <td><p>A vector of optimal amino acids for each site to be simulated.</p></td>
    </tr>
    <tr>
      <th>root.codon.frequencies</th>
      <td><p>A vector of codon frequencies for each possible optimal amino acid. Thus, the vector is of length 64x64.</p></td>
    </tr>
    <tr>
      <th>numcode</th>
      <td><p>The The ncbi genetic code number for translation. By default the standard (numcode=1) genetic code is used.</p></td>
    </tr>
    <tr>
      <th>aa.properties</th>
      <td><p>User-supplied amino acid distance properties. By default we assume Grantham (1974) properties.</p></td>
    </tr>
    <tr>
      <th>nuc.model</th>
      <td><p>Indicates what type nucleotide model to use. There are three options: "JC", "GTR", or "UNREST".</p></td>
    </tr>
    <tr>
      <th>k.levels</th>
      <td><p>Provides how many levels in the polynomial. By default we assume a single level (i.e., linear).</p></td>
    </tr>
    <tr>
      <th>diploid</th>
      <td><p>A logical indicating whether or not the organism is diploid or not.</p></td>
    </tr>
    <tr>
      <th>pars.to.evolve</th>
      <td><p>Indicates which parameters to assume evolve along the tree. Only two options: "phi" or "Ne".</p></td>
    </tr>
    <tr>
      <th>evolve.type</th>
      <td><p>The process by which the focal parameter evovles. There are two options: Brownian motion ("BM") or Ornstein-Uhlenbeck ("OU").</p></td>
    </tr>
    <tr>
      <th>evolve.pars</th>
      <td><p>The process parameters used to simulate focal parameter evolution. Under "BM", the order is root.state, rate; under "OU", the order is alpha, sigma.sq, and the mean.</p></td>
    </tr>
    <tr>
      <th>Ne.vals.evolved</th>
      <td><p>Under selac we assume a global Ne for all genes. Thus, when the focal parameter to evolve is "Ne", then a user specified vector of simulated Ne values are provided here.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Simulates a nucleotide matrix using parameters under the SELAC model, but allows either Phi or Ne to evolve along the tree. Note that the output can be written to a fasta file using the write.dna() function in the <code>ape</code> package.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#details">Details</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Jeremy Beaulieu, JJ Chai, Brian O'Meara.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
