<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Efficient optimization of the SELON model — SelonLargeOptimize • selac</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">selac</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Efficient optimization of the SELON model</h1>
    </div>

    
    <p>Efficient optimization of model parameters under the SELON model</p>
    

    <pre><span class='fu'>SelonLargeOptimize</span>(<span class='no'>nuc.data.path</span>, <span class='kw'>n.partitions</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>phy</span>,
  <span class='kw'>edge.length</span> <span class='kw'>=</span> <span class='st'>"optimize"</span>, <span class='kw'>edge.linked</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>optimal.nuc</span> <span class='kw'>=</span> <span class='st'>"majrule"</span>,
  <span class='kw'>nuc.model</span> <span class='kw'>=</span> <span class='st'>"GTR"</span>, <span class='kw'>diploid</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>n.cores</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>max.tol</span> <span class='kw'>=</span> <span class='no'>.Machine</span>$<span class='no'>double.eps</span>^<span class='fl'>0.5</span>, <span class='kw'>max.evals</span> <span class='kw'>=</span> <span class='fl'>1e+06</span>, <span class='kw'>max.restarts</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>output.by.restart</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>output.restart.filename</span> <span class='kw'>=</span> <span class='st'>"restartResult"</span>,
  <span class='kw'>fasta.rows.to.keep</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>nuc.data.path</th>
      <td><p>Provides the path to the directory containing the gene specific fasta files that contains the nucleotide data.</p></td>
    </tr>
    <tr>
      <th>n.partitions</th>
      <td><p>The number of partitions to analyze. The order is based on the Unix order of the fasta files in the directory.</p></td>
    </tr>
    <tr>
      <th>phy</th>
      <td><p>The phylogenetic tree to optimize the model parameters.</p></td>
    </tr>
    <tr>
      <th>edge.length</th>
      <td><p>Indicates whether or not edge lengths should be optimized. By default it is set to "optimize", other option is "fixed", which user-supplied branch lengths.</p></td>
    </tr>
    <tr>
      <th>edge.linked</th>
      <td><p>A logical indicating whether or not edge lengths should be optimized separately for each gene. By default, a single set of each lengths is optimized for all genes.</p></td>
    </tr>
    <tr>
      <th>optimal.nuc</th>
      <td><p>Indicates what type of optimal.nuc should be used. At the moment there is only a single option: "majrule".</p></td>
    </tr>
    <tr>
      <th>nuc.model</th>
      <td><p>Indicates what type nucleotide model to use. There are three options: "JC", "GTR", or "UNREST".</p></td>
    </tr>
    <tr>
      <th>diploid</th>
      <td><p>A logical indicating whether or not the organism is diploid or not.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical indicating whether each iteration be printed to the screen.</p></td>
    </tr>
    <tr>
      <th>n.cores</th>
      <td><p>The number of cores to run the analyses over.</p></td>
    </tr>
    <tr>
      <th>max.tol</th>
      <td><p>Supplies the relative optimization tolerance.</p></td>
    </tr>
    <tr>
      <th>max.evals</th>
      <td><p>Supplies the max number of iterations tried during optimization.</p></td>
    </tr>
    <tr>
      <th>max.restarts</th>
      <td><p>Supplies the number of random restarts.</p></td>
    </tr>
    <tr>
      <th>output.by.restart</th>
      <td><p>Logical indicating whether or not each restart is saved to a file. Default is TRUE.</p></td>
    </tr>
    <tr>
      <th>output.restart.filename</th>
      <td><p>Designates the file name for each random restart.</p></td>
    </tr>
    <tr>
      <th>fasta.rows.to.keep</th>
      <td><p>Indicates which rows to remove in the input fasta files.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This is the exact same implementation as in SelonOptimize(), except here we optimize parameters across each gene separately while keeping the shared parameters, the edge lengths, constant across genes. We then optimize edge lengths while keeping parameters for each gene fixed. This approach is potentially more efficient than simply optimizing all parameters simultaneously, especially if fitting models across 100's of genes.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#details">Details</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Jeremy Beaulieu, JJ Chai, Brian O'Meara.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
